### PR TITLE
Add FileOpener interface in Hive connector

### PR DESCRIPTION
**Pull request by shixuan-fan**
```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/FileOpener.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.io.IOException;
+import java.util.Optional;
+
+public interface FileOpener
+{
+    FSDataInputStream open(FileSystem fileSystem, Path path, Optional<byte[]> extraFileInfo)
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/FileOpener.java]: nit: I wonder if the `Optional<byte[]> fileDescriptor` would be a better name. If we decided to change the name so - please change it everywhere.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveFileInfo.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import static java.util.Objects.requireNonNull;
+
+public class HiveFileInfo
+        implements Comparable
+{
+    private final Path path;
+    private final boolean isDirectory;
+    private final BlockLocation[] blockLocations;
+    private final long length;
+    private final Optional<byte[]> extraFileContext;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveFileInfo.java]: Lets be consistent with naming

**Comment by shixuan-fan** [comment]
As for naming, my initial thought was also `fileDescriptor`, but @wenleix mentioned that file descriptor has its own meaning (https://www.computerhope.com/jargon/f/file-descriptor.htm), which might be a little bit confusing if used here.

**Comment by wenleix** [comment]
@arhimondr :

> The only question i have is whether we should rename extraFileContext/extraFileInfo to be fileDescriptor.

@shixuan-fan and I discussed this before. I personally feel `FileDescriptor` makes me feel like a `FileHandler` so I suggest to use something like `extraFileContext/Info` :)

File: presto-hive/src/main/java/com/facebook/presto/hive/InternalHiveSplit.java [hunk 3]
```diff
@@ -62,6 +62,7 @@
     private final boolean forceLocalScheduling;
     private final boolean s3SelectPushdownEnabled;
     private final HiveSplitPartitionInfo partitionInfo;
+    private final Optional<byte[]> extraFileInfo;
```
Comment (arhimondr) [inline presto-hive/src/main/java/com/facebook/presto/hive/InternalHiveSplit.java]: Please update memory accounting code

Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/InternalHiveSplit.java]: Done.

**Comment by wenleix** [comment]
Looks good.

**Pull request merged**